### PR TITLE
Add Tesseract training setup scripts and example data

### PR DESCRIPTION
**Pull request by Penguin2600**
Work In progress, opening for visibility.

Current status:

tessTrain/tessTrain.sh - works and will set up a baseline ubuntu 22.04 wsl / container / etc with the tools and binaries required for training tesseract. It will also run an example training session with the included example training data. Documentation and sources are commented inside the script for further details look there for now.

tessTrain/example_truth/ - Example of what a training data directory needs to look like. Used by tessTrain.sh to confirm that setup was successful.

Ping me on discord for any questions or comments! Thx.

### CODE REVIEW

File: dataScripts/tessTrain/tessTrain.sh [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+# Dont polute the directory
+mkdir -p ./tess
+cd tess
+pwd
+
+# Get Deps
+sudo apt-get install libicu-dev libpango1.0-dev libcairo2-dev
+sudo apt-get install automake ca-certificates g++ git libtool libleptonica-dev make pkg-config
+sudo apt-get install libpango1.0-dev libleptonica-dev
```
Comment (TwoAbove) [inline dataScripts/tessTrain/tessTrain.sh:27]: I think it would make sense to extract this into the README as a `## training tesseract` section. 

I would split this script into two parts - a `setup.sh` script (also mention it in the README in the setup instructions) and a `train.sh` script that takes in a ground truth path.

File: dataScripts/tessTrain/tessTrain.sh [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+greentext () {
+   echo -e "\n${BGreen}### $1 ${NC}\n"
+}
+
+# Build tesseract
+# Upstream Docs: https://tesseract-ocr.github.io/tessdoc/Compiling-%E2%80%93-GitInstallation#installing-with-autoconf-tools
+greentext "Installing Deps and Creating File Structure"
+
+# Dont polute the directory
+mkdir -p ./tess
```
Comment (TwoAbove) [inline dataScripts/tessTrain/tessTrain.sh:20]: Since this script creates artifacts, we'll need to add them to a `.gitignore` file. Ideally, we would keep the sole `.gitignore` so it's consolidated in one place.

File: dataScripts/tessTrain/tessTrain.sh [hunk 3]
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+set -e
+
+# Tested on a bare install of Ubuntu 22.04.3 LTS - 02/27/2024
+# Author: Charles@CharlesBock.com - github.com/Penguin2600
```
Comment (TwoAbove) [inline dataScripts/tessTrain/tessTrain.sh:5]: If you want to add author info here, (if I recall correctly), it should go right after the shebang. Do correct me if I'm wrong!

```suggestion
#!/bin/bash
# Author: Charles Bock
# Email: Charles@CharlesBock.com
# Date: 2024-02-27
# Tested on a bare install of Ubuntu 22.04.3 LTS

set -e
```

File: dataScripts/tessTrain/tessTrain.sh [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+git clone https://github.com/tesseract-ocr/tesseract.git 2> /dev/null || (cd tesseract ; git pull; cd ..)
+
+# Build the training tools
+# Upstream Docs: https://tesseract-ocr.github.io/tessdoc/Compiling-%E2%80%93-GitInstallation#build-with-training-tools
+greentext "Building Tesseract WITH Training Tools - This can take a long time"
+
+sudo apt-get install libicu-dev
+sudo apt-get install libpango1.0-dev
+sudo apt-get install libcairo2-dev
```
Comment (TwoAbove) [inline dataScripts/tessTrain/tessTrain.sh:39]: These too

File: dataScripts/tessTrain/tessTrain.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+cd ..
+pwd
+
+# Install and configure tesstrain
+# Upstream Docs: https://github.com/tesseract-ocr/tesstrain?tab=readme-ov-file#choose-model-name
+
+greentext "Pulling the required ENG traineddata from github"
+wget https://github.com/tesseract-ocr/tessdata/raw/main/eng.traineddata
+sudo mv eng.traineddata /usr/local/share/tessdata
```
Comment (TwoAbove) [inline dataScripts/tessTrain/tessTrain.sh:61]: Is there a way to not populate paths outside of noitool? It would be great if this would be confined to this directory. It looks like you can use `TESSDATA` variable to make it local. https://github.com/tesseract-ocr/tesstrain?tab=readme-ov-file#train

**Review by TwoAbove — COMMENTED** [review]
Looks good! 

Left a couple of minor comments.

I also have a question about dataScripts/tessTrain/example_truth/97984949.png and similar ones. Would the extra M hinder the training in any way?

File: dataScripts/tessTrain/tessTrain.sh (hunk 1 above)
Comment (Penguin2600) [reply dataScripts/tessTrain/tessTrain.sh:27]: Totally, I had the same thought. one will likely run once while the other may need many runs.

File: dataScripts/tessTrain/tessTrain.sh (hunk 2 above)
Comment (Penguin2600) [reply dataScripts/tessTrain/tessTrain.sh:20]: Good Callout, I'll consider that the new entries might need to be.

File: dataScripts/tessTrain/tessTrain.sh (hunk 3 above)
Comment (Penguin2600) [reply dataScripts/tessTrain/tessTrain.sh:5]: I'm good for whatever you like, Just wanted to make sure if some poor sap runs into this in the future they know who they can contact for help.

File: dataScripts/tessTrain/tessTrain.sh (hunk 5 above)
Comment (Penguin2600) [reply dataScripts/tessTrain/tessTrain.sh:61]: Yes, I think that's a great idea, will incorporate.

**Comment by Penguin2600** [comment]
> I also have a question about dataScripts/tessTrain/example_truth/97984949.png and similar ones. Would the extra M hinder the training in any way?

Short Answer: 
I don't know. 

Long answer: 
My best understanding is that if it tags that partial M as a glyph (box generation step) then it may, in the training, try and label it. Which it will almost certainly fail at since it is not present in the corresponding gt.txt file. This will increase the resulting training error due the mismatch which may in a sort of artificial way cause that training data to be ignored in the training. Ideally we should give it perfectly clean images and a perfectly clean matching ground_truth text.  In this case depending on the training execution variables I see 90-98% success rates on validation data out of the minimal example training set.
